File: sdk/java-holons/src/test/java/org/organicprogramming/holons/ObservabilityTest.java
package org.organicprogramming.holons;

import io.grpc.CallOptions;
import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import io.grpc.Server;
import io.grpc.ServerBuilder;
import io.grpc.stub.ClientCalls;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.TimeUnit;

import static org.junit.jupiter.api.Assertions.*;

final class ObservabilityTest {
    @Test
    void parseOpObsRejectsV2Tokens() {
        Set<Observability.Family> all = Set.of(
                Observability.Family.LOGS,
                Observability.Family.METRICS,
                Observability.Family.EVENTS,
                Observability.Family.PROM);
        assertEquals(all, Observability.parseOpObs("all"));
        assertThrows(Observability.InvalidTokenException.class, () -> Observability.parseOpObs("all,otel"));
        assertThrows(Observability.InvalidTokenException.class, () -> Observability.parseOpObs("all,sessions"));
        assertThrows(Observability.InvalidTokenException.class, () -> Observability.parseOpObs("unknown"));
    }

    @Test
    void checkEnvRejectsV2TokensAndOpSessions() {
        assertThrows(Observability.InvalidTokenException.class,
                () -> Observability.checkEnv(Map.of("OP_OBS", "logs,otel")));
        assertThrows(Observability.InvalidTokenException.class,
                () -> Observability.checkEnv(Map.of("OP_OBS", "logs,sessions")));
        Observability.InvalidTokenException err = assertThrows(Observability.InvalidTokenException.class,
                () -> Observability.checkEnv(Map.of("OP_SESSIONS", "metrics")));
        assertEquals("OP_SESSIONS", err.variable);
        assertDoesNotThrow(() -> Observability.checkEnv(Map.of("OP_OBS", "logs,metrics,events,prom,all")));
    }

    @Test
    void runDirDerivesFromRegistryRoot(@TempDir Path tmp) {
        Observability.reset();
        Observability.Config cfg = new Observability.Config();
        cfg.slug = "gabriel-greeting-java";
        cfg.instanceUid = "uid-1";
        cfg.runDir = tmp.toString();

        Observability obs = Observability.configureFromEnv(cfg, Map.of("OP_OBS", "logs"));

        assertEquals(tmp.resolve("gabriel-greeting-java").resolve("uid-1").toString(), obs.cfg.runDir);
        Observability.reset();
    }

    @Test
    void serviceReplaysLogsMetricsEventsAndDiskFiles(@TempDir Path tmp) throws Exception {
        Observability.reset();
        Observability.Config cfg = new Observability.Config();
        cfg.slug = "gabriel-greeting-java";
        cfg.instanceUid = "uid-2";
        cfg.runDir = tmp.toString();
        Observability obs = Observability.configureFromEnv(
                cfg,
                Map.of("OP_OBS", "logs,metrics,events"));

        Observability.enableDiskWriters(obs.cfg.runDir);
        obs.logger("test").info("service-log", Map.of("component", "java"));
        obs.counter("java_requests_total", "", Map.of()).inc();
        obs.emit(Observability.EVENT_INSTANCE_READY, Map.of("listener", "tcp://127.0.0.1:1"));
        Observability.MetaJson meta = new Observability.MetaJson();
        meta.slug = obs.cfg.slug;
        meta.uid = obs.cfg.instanceUid;
        meta.pid = 123;
        meta.transport = "tcp";
        meta.address = "tcp://127.0.0.1:1";
        meta.logPath = Path.of(obs.cfg.runDir, "stdout.log").toString();
        Observability.writeMetaJson(obs.cfg.runDir, meta);

        Path runDir = tmp.resolve("gabriel-greeting-java").resolve("uid-2");
        assertTrue(Files.readString(runDir.resolve("stdout.log")).contains("\"message\":\"service-log\""));
        assertTrue(Files.readString(runDir.resolve("events.jsonl")).contains("\"event_name\":\"instance.ready\""));
        assertTrue(Files.readString(runDir.resolve("meta.json")).contains("\"uid\":\"uid-2\""));

        Server server = ServerBuilder.forPort(0)
                .addService(Observability.service(obs))
                .build()
                .start();
        ManagedChannel channel = ManagedChannelBuilder.forAddress("127.0.0.1", server.getPort())
                .usePlaintext()
                .build();
        try {
            List<holons.v1.Observability.LogRecord> logs = new ArrayList<>();
            Iterator<holons.v1.Observability.LogRecord> logIterator = ClientCalls.blockingServerStreamingCall(
                    channel,
                    Observability.logsMethod(),
                    CallOptions.DEFAULT,
                    holons.v1.Observability.LogsRequest.newBuilder()
                            .setMinSeverityNumber(holons.v1.Observability.SeverityNumber.SEVERITY_NUMBER_INFO)
                            .build());
            logIterator.forEachRemaining(logs::add);
            assertTrue(logs.stream().anyMatch(entry -> entry.getBody().getStringValue().equals("service-log")));

            List<holons.v1.Observability.Metric> metrics = new ArrayList<>();
            Iterator<holons.v1.Observability.Metric> metricIterator = ClientCalls.blockingServerStreamingCall(
                    channel,
                    Observability.metricsMethod(),
                    CallOptions.DEFAULT,
                    holons.v1.Observability.MetricsRequest.getDefaultInstance());
            metricIterator.forEachRemaining(metrics::add);
            assertTrue(metrics.stream().anyMatch(metric -> metric.getName().equals("java_requests_total")
                    && metric.hasSum()
                    && metric.getSum().getIsMonotonic()
                    && metric.getSum().getAggregationTemporality()
                    == holons.v1.Observability.AggregationTemporality.AGGREGATION_TEMPORALITY_CUMULATIVE));

            List<holons.v1.Observability.LogRecord> events = new ArrayList<>();
            Iterator<holons.v1.Observability.LogRecord> eventIterator = ClientCalls.blockingServerStreamingCall(
                    channel,
                    Observability.eventsMethod(),
                    CallOptions.DEFAULT,
                    holons.v1.Observability.EventsRequest.getDefaultInstance());
            eventIterator.forEachRemaining(events::add);
            assertTrue(events.stream()
                    .anyMatch(event -> event.getEventName().equals(Observability.EVENT_INSTANCE_READY)));
        } finally {
            channel.shutdownNow();
            channel.awaitTermination(5, TimeUnit.SECONDS);
            server.shutdownNow();
            server.awaitTermination(5, TimeUnit.SECONDS);
            Observability.reset();
        }
    }

    @Test
    void prometheusTextInjectsIdentityLabelsAndRoundTripsProto() {
        Observability.reset();
        Observability.Config cfg = new Observability.Config();
        cfg.slug = "cascade-node-java";
        cfg.instanceUid = "java-prom-1";
        Observability obs = Observability.configureFromEnv(cfg, Map.of("OP_OBS", "logs,metrics,events,prom"));

        obs.counter("cascade_ticks_total", "Ticks received by this cascade node.", Map.of("responder_uid", "java-prom-1")).inc();
        String text = Observability.toPrometheusText(obs);
        assertTrue(text.contains("# HELP cascade_ticks_total Ticks received by this cascade node."));
        assertTrue(text.contains("# TYPE cascade_ticks_total counter"));
        assertTrue(text.contains("instance_uid=\"java-prom-1\""));
        assertTrue(text.contains("responder_uid=\"java-prom-1\""));
        assertTrue(text.contains("slug=\"cascade-node-java\""));

        Observability.LogRecord log = new Observability.LogRecord(holons.v1.Observability.LogRecord.newBuilder()
                .setBody(Observability.anyValue("tick received"))
                .addChain("leaf")
                .addAttributes(holons.v1.Observability.KeyValue.newBuilder()
                        .setKey("sender")
                        .setValue(Observability.anyValue("test")))
                .build());
        Observability.LogRecord roundTrip = Observability.fromProtoLogRecord(Observability.toProtoLogRecord(log));
        assertEquals("tick received", roundTrip.bodyString());
        assertEquals("leaf", roundTrip.record.getChain(0));
        Observability.reset();
    }

    @Test
    void testLogsFollowReplaysRingOnSubscribe() throws Exception {
        Observability.reset();
        Observability.Config cfg = new Observability.Config();
        cfg.slug = "java-log-replay";
        cfg.instanceUid = "log-replay-1";
        Observability obs = Observability.configureFromEnv(cfg, Map.of("OP_OBS", "logs"));
        obs.logger("test").info("before-subscribe", Map.of());

        Server server = ServerBuilder.forPort(0)
                .addService(Observability.service(obs))
                .build()
                .start();
        ManagedChannel channel = ManagedChannelBuilder.forAddress("127.0.0.1", server.getPort())
                .usePlaintext()
                .build();
        try {
            Iterator<holons.v1.Observability.LogRecord> iterator = ClientCalls.blockingServerStreamingCall(
                    channel,
                    Observability.logsMethod(),
                    CallOptions.DEFAULT.withDeadlineAfter(2, TimeUnit.SECONDS),
                    holons.v1.Observability.LogsRequest.newBuilder()
                            .setFollow(true)
                            .setMinSeverityNumber(holons.v1.Observability.SeverityNumber.SEVERITY_NUMBER_INFO)
                            .build());

            assertTrue(iterator.hasNext());
            assertEquals("before-subscribe", iterator.next().getBody().getStringValue());
            obs.logger("test").info("after-subscribe", Map.of());
            assertTrue(iterator.hasNext());
            assertEquals("after-subscribe", iterator.next().getBody().getStringValue());
        } finally {
            channel.shutdownNow();
            channel.awaitTermination(5, TimeUnit.SECONDS);
            server.shutdownNow();
            server.awaitTermination(5, TimeUnit.SECONDS);
            Observability.reset();
        }
    }

    @Test
    void testEventsFollowReplaysRingOnSubscribe() throws Exception {
        Observability.reset();
        Observability.Config cfg = new Observability.Config();
        cfg.slug = "java-event-replay";
        cfg.instanceUid = "event-replay-1";
        Observability obs = Observability.configureFromEnv(cfg, Map.of("OP_OBS", "events"));
        obs.emit(Observability.EVENT_INSTANCE_READY, Map.of("listener", "tcp://127.0.0.1:1"));

        Server server = ServerBuilder.forPort(0)
                .addService(Observability.service(obs))
                .build()
                .start();
        ManagedChannel channel = ManagedChannelBuilder.forAddress("127.0.0.1", server.getPort())
                .usePlaintext()
                .build();
        try {
            Iterator<holons.v1.Observability.LogRecord> iterator = ClientCalls.blockingServerStreamingCall(
                    channel,
                    Observability.eventsMethod(),
                    CallOptions.DEFAULT.withDeadlineAfter(2, TimeUnit.SECONDS),
                    holons.v1.Observability.EventsRequest.newBuilder()
                            .setFollow(true)
                            .build());

            assertTrue(iterator.hasNext());
            assertEquals(Observability.EVENT_INSTANCE_READY, iterator.next().getEventName());
            obs.emit(Observability.EVENT_INSTANCE_EXITED, Map.of("listener", "tcp://127.0.0.1:1"));
            assertTrue(iterator.hasNext());
            assertEquals(Observability.EVENT_INSTANCE_EXITED, iterator.next().getEventName());
        } finally {
            channel.shutdownNow();
            channel.awaitTermination(5, TimeUnit.SECONDS);
            server.shutdownNow();
            server.awaitTermination(5, TimeUnit.SECONDS);
            Observability.reset();
        }
    }
}

File: sdk/kotlin-holons/src/test/kotlin/org/organicprogramming/holons/ObservabilityTest.kt
package org.organicprogramming.holons

import holons.v1.Observability as ObsProto
import io.grpc.CallOptions
import io.grpc.ManagedChannelBuilder
import io.grpc.ServerBuilder
import io.grpc.stub.ClientCalls
import java.nio.file.Files
import java.nio.file.Path
import java.util.concurrent.TimeUnit
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertTrue

class ObservabilityTest {
    @Test
    fun parseOpObsRejectsV2Tokens() {
        val all = setOf(
            Observability.Family.LOGS,
            Observability.Family.METRICS,
            Observability.Family.EVENTS,
            Observability.Family.PROM,
        )
        assertEquals(all, Observability.parseOpObs("all"))
        assertFailsWith<Observability.InvalidTokenException> {
            Observability.parseOpObs("all,otel")
        }
        assertFailsWith<Observability.InvalidTokenException> {
            Observability.parseOpObs("all,sessions")
        }
        assertFailsWith<Observability.InvalidTokenException> {
            Observability.parseOpObs("unknown")
        }
    }

    @Test
    fun checkEnvRejectsV2TokensAndOpSessions() {
        assertFailsWith<Observability.InvalidTokenException> {
            Observability.checkEnv(mapOf("OP_OBS" to "logs,otel"))
        }
        assertFailsWith<Observability.InvalidTokenException> {
            Observability.checkEnv(mapOf("OP_OBS" to "logs,sessions"))
        }
        val err = assertFailsWith<Observability.InvalidTokenException> {
            Observability.checkEnv(mapOf("OP_SESSIONS" to "metrics"))
        }
        assertEquals("OP_SESSIONS", err.variable)
    }

    @Test
    fun runDirDerivesFromRegistryRoot() {
        Observability.reset()
        val root = Files.createTempDirectory("kotlin-obs-root")
        try {
            val inst = Observability.configureFromEnv(
                Observability.Config(
                    slug = "gabriel-greeting-kotlin",
                    instanceUid = "uid-1",
                    runDir = root.toString(),
                ),
                mapOf("OP_OBS" to "logs"),
            )

            assertEquals(root.resolve("gabriel-greeting-kotlin").resolve("uid-1").toString(), inst.cfg.runDir)
        } finally {
            Observability.reset()
            root.toFile().deleteRecursively()
        }
    }

    @Test
    fun serviceReplaysLogsMetricsEventsAndDiskFiles() {
        Observability.reset()
        val root = Files.createTempDirectory("kotlin-obs-service")
        try {
            val inst = Observability.configureFromEnv(
                Observability.Config(
                    slug = "gabriel-greeting-kotlin",
                    instanceUid = "uid-2",
                    runDir = root.toString(),
                ),
                mapOf("OP_OBS" to "logs,metrics,events"),
            )
            Observability.enableDiskWriters(inst.cfg.runDir)
            inst.logger("test").info("service-log", mapOf("component" to "kotlin"))
            inst.counter("kotlin_requests_total")!!.inc()
            inst.emit(Observability.EventName.INSTANCE_READY, mapOf("listener" to "tcp://127.0.0.1:1"))
            Observability.writeMetaJson(
                inst.cfg.runDir,
                Observability.MetaJson(
                    slug = inst.cfg.slug,
                    uid = inst.cfg.instanceUid,
                    pid = 123,
                    transport = "tcp",
                    address = "tcp://127.0.0.1:1",
                    logPath = Path.of(inst.cfg.runDir, "stdout.log").toString(),
                ),
            )

            val runDir = root.resolve("gabriel-greeting-kotlin").resolve("uid-2")
            assertTrue(Files.readString(runDir.resolve("stdout.log")).contains("\"message\":\"service-log\""))
            assertTrue(Files.readString(runDir.resolve("events.jsonl")).contains("\"event_name\":\"instance.ready\""))
            assertTrue(Files.readString(runDir.resolve("meta.json")).contains("\"uid\":\"uid-2\""))

            val server = ServerBuilder.forPort(0)
                .addService(Observability.service(inst))
                .build()
                .start()
            val channel = ManagedChannelBuilder.forAddress("127.0.0.1", server.port).usePlaintext().build()
            try {
                val logs = ClientCalls.blockingServerStreamingCall(
                    channel,
                    Observability.logsMethod,
                    CallOptions.DEFAULT,
                    ObsProto.LogsRequest.newBuilder()
                        .setMinSeverityNumber(ObsProto.SeverityNumber.SEVERITY_NUMBER_INFO)
                        .build(),
                ).asSequence().toList()
                assertTrue(logs.any { it.body.stringValue == "service-log" })

                val metrics = ClientCalls.blockingServerStreamingCall(
                    channel,
                    Observability.metricsMethod,
                    CallOptions.DEFAULT,
                    ObsProto.MetricsRequest.getDefaultInstance(),
                ).asSequence().toList()
                assertTrue(metrics.any {
                    it.name == "kotlin_requests_total" &&
                        it.hasSum() &&
                        it.sum.isMonotonic &&
                        it.sum.aggregationTemporality ==
                        ObsProto.AggregationTemporality.AGGREGATION_TEMPORALITY_CUMULATIVE
                })

                val events = ClientCalls.blockingServerStreamingCall(
                    channel,
                    Observability.eventsMethod,
                    CallOptions.DEFAULT,
                    ObsProto.EventsRequest.getDefaultInstance(),
                ).asSequence().toList()
                assertTrue(events.any { it.eventName == Observability.EventName.INSTANCE_READY })
            } finally {
                channel.shutdownNow()
                channel.awaitTermination(5, TimeUnit.SECONDS)
                server.shutdownNow()
                server.awaitTermination(5, TimeUnit.SECONDS)
            }
        } finally {
            Observability.reset()
            root.toFile().deleteRecursively()
        }
    }

    @Test
    fun prometheusTextInjectsIdentityLabelsAndRoundTripsProto() {
        Observability.reset()
        try {
            val inst = Observability.configureFromEnv(
                Observability.Config(
                    slug = "cascade-node-kotlin",
                    instanceUid = "kotlin-prom-1",
                ),
                mapOf("OP_OBS" to "logs,metrics,events,prom"),
            )
            inst.counter(
                "cascade_ticks_total",
                "Ticks received by this cascade node.",
                mapOf("responder_uid" to "kotlin-prom-1"),
            )!!.inc()

            val text = Observability.toPrometheusText(inst)
            assertTrue(text.contains("# HELP cascade_ticks_total Ticks received by this cascade node."))
            assertTrue(text.contains("# TYPE cascade_ticks_total counter"))
            assertTrue(text.contains("instance_uid=\"kotlin-prom-1\""))
            assertTrue(text.contains("responder_uid=\"kotlin-prom-1\""))
            assertTrue(text.contains("slug=\"cascade-node-kotlin\""))

            val log = Observability.LogRecord(
                timestamp = java.time.Instant.now(),
                level = Observability.Level.INFO,
                slug = "child",
                instanceUid = "uid-child",
                message = "tick received",
                fields = mapOf("sender" to "test"),
                chain = listOf("leaf"),
            )
            val roundTrip = Observability.fromProtoLogRecord(Observability.toProtoLogRecord(log))
            assertEquals("tick received", roundTrip.message)
            assertEquals("leaf", roundTrip.chain.single())
        } finally {
            Observability.reset()
        }
    }

    @Test
    fun testLogsFollowReplaysRingOnSubscribe() {
        Observability.reset()
        val inst = Observability.configureFromEnv(
            Observability.Config(slug = "kotlin-log-replay", instanceUid = "log-replay-1"),
            mapOf("OP_OBS" to "logs"),
        )
        inst.logger("test").info("before-subscribe")

        val server = ServerBuilder.forPort(0)
            .addService(Observability.service(inst))
            .build()
            .start()
        val channel = ManagedChannelBuilder.forAddress("127.0.0.1", server.port).usePlaintext().build()
        try {
            val iterator = ClientCalls.blockingServerStreamingCall(
                channel,
                Observability.logsMethod,
                CallOptions.DEFAULT.withDeadlineAfter(2, TimeUnit.SECONDS),
                ObsProto.LogsRequest.newBuilder()
                    .setFollow(true)
                    .setMinSeverityNumber(ObsProto.SeverityNumber.SEVERITY_NUMBER_INFO)
                    .build(),
            )

            assertTrue(iterator.hasNext())
            assertEquals("before-subscribe", iterator.next().body.stringValue)
            inst.logger("test").info("after-subscribe")
            assertTrue(iterator.hasNext())
            assertEquals("after-subscribe", iterator.next().body.stringValue)
        } finally {
            channel.shutdownNow()
            channel.awaitTermination(5, TimeUnit.SECONDS)
            server.shutdownNow()
            server.awaitTermination(5, TimeUnit.SECONDS)
            Observability.reset()
        }
    }

    @Test
    fun testEventsFollowReplaysRingOnSubscribe() {
        Observability.reset()
        val inst = Observability.configureFromEnv(
            Observability.Config(slug = "kotlin-event-replay", instanceUid = "event-replay-1"),
            mapOf("OP_OBS" to "events"),
        )
        inst.emit(Observability.EventName.INSTANCE_READY, mapOf("listener" to "tcp://127.0.0.1:1"))

        val server = ServerBuilder.forPort(0)
            .addService(Observability.service(inst))
            .build()
            .start()
        val channel = ManagedChannelBuilder.forAddress("127.0.0.1", server.port).usePlaintext().build()
        try {
            val iterator = ClientCalls.blockingServerStreamingCall(
                channel,
                Observability.eventsMethod,
                CallOptions.DEFAULT.withDeadlineAfter(2, TimeUnit.SECONDS),
                ObsProto.EventsRequest.newBuilder()
                    .setFollow(true)
                    .build(),
            )

            assertTrue(iterator.hasNext())
            assertEquals(Observability.EventName.INSTANCE_READY, iterator.next().eventName)
            inst.emit(Observability.EventName.INSTANCE_EXITED, mapOf("listener" to "tcp://127.0.0.1:1"))
            assertTrue(iterator.hasNext())
            assertEquals(Observability.EventName.INSTANCE_EXITED, iterator.next().eventName)
        } finally {
            channel.shutdownNow()
            channel.awaitTermination(5, TimeUnit.SECONDS)
            server.shutdownNow()
            server.awaitTermination(5, TimeUnit.SECONDS)
            Observability.reset()
        }
    }
}
